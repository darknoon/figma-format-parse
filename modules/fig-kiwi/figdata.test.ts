import FigmaArchiveParser from "./archive";
import FigDataParser from "./figdata";
import { readFileSync, writeFileSync } from "fs";
import {
  compileSchema,
  decodeBinarySchema,
  compileSchemaTypeScript,
  prettyPrintSchema,
} from "kiwi-schema";
import schema from "./schema";
import { Schema as CompiledSchema, NodeChange, Message } from "./schema-defs";
import { inflateRaw } from "pako";

test("this just formats the schema", () => {
  const prettySchema = prettyPrintSchema(schema);
  writeFileSync(__dirname + "/figma-pretty.kiwi", prettySchema);
});

function parseFile(data: Uint8Array): Message {
  const parsed = FigmaArchiveParser.parseArchive(data);
  const [schemaFile, dataFile, previewFile] = parsed.files;
  const fileSchema = decodeBinarySchema(inflateRaw(schemaFile));
  const parser = new FigDataParser(inflateRaw(dataFile), fileSchema);
  return parser.parseMessage();
}

test("able to parse figma kiwi", () => {
  const data = readFileSync(__dirname + "/data/blue-circle.fig");
  const parsed = FigmaArchiveParser.parseArchive(data);
  const [schemaFile, dataFile, previewFile] = parsed.files;
  const figDataDec = inflateRaw(dataFile);
  expect(figDataDec.length).toBeGreaterThan(100);
  writeFileSync(__dirname + "/data/blue-circle.fig-inflated", figDataDec);
  const fileSchema = decodeBinarySchema(inflateRaw(schemaFile));
  expect(fileSchema).toHaveProperty("definitions");
  const parser = new FigDataParser(figDataDec, fileSchema);
  const decoded = parser.parseMessage();
  expect(decoded).not.toBeNull();
});

test("compare red and blue", () => {
  const blue = parseFile(readFileSync(__dirname + "/data/blue-circle.fig"));
  const red = parseFile(readFileSync(__dirname + "/data/red-circle.fig"));
  expect(blue).toMatchSnapshot();
});

test("able to enc dec a known message", () => {
  const cs = compileSchema(schema) as CompiledSchema;
  const message: NodeChange = { guid: { sessionID: 123, localID: 321 } };
  const data = cs.encodeNodeChange(message);
  const decoded = cs.decodeNodeChange(data);
  expect(decoded).toEqual(message);
});

test("it's compressed binary kiwi format in the metadata", () => {
  const meta = readFileSync(__dirname + "/data/red-circle.fig-meta");
  const schemaBin = inflateRaw(meta);
  const figMetaSchema = decodeBinarySchema(schemaBin);
  writeFileSync(
    __dirname + "/data/red-circle.kiwi",
    prettyPrintSchema(figMetaSchema)
  );

  // writeFileSync(
  //   __dirname + "/fig-kiwi.ts",
  //   compileSchemaTypeScript(figMetaSchema)
  // );
  // console.log(tsdef);
});

test("able to decode a real websocket message", () => {
  const cs = compileSchema(schema) as CompiledSchema;
  const dataCompressed = readFileSync(__dirname + "/data/set-black.bin");
  const data = inflateRaw(dataCompressed);
  const decoded = cs.decodeMessage(data);
  expect(decoded).not.toBeNull();
  expect(decoded).toMatchSnapshot();
});
