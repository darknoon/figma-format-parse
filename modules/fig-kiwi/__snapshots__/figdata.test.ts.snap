// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`able to decode a real websocket message 1`] = `
Object {
  "ackID": 265,
  "nodeChanges": Array [
    Object {
      "fillPaints": Array [
        Object {
          "blendMode": "NORMAL",
          "color": Object {
            "a": 1,
            "b": 0,
            "g": 0,
            "r": 0,
          },
          "opacity": 1,
          "type": "SOLID",
          "visible": true,
        },
      ],
      "guid": Object {
        "localID": 5,
        "sessionID": 9,
      },
    },
  ],
  "sessionID": 9,
  "type": "NODE_CHANGES",
}
`;

exports[`compare red and blue 1`] = `
Object {
  "ackID": 0,
  "blobs": Array [
    Object {
      "bytes": Uint8Array [
        1,
        0,
        0,
        17,
        67,
        0,
        0,
        145,
        66,
        4,
        0,
        0,
        17,
        67,
        207,
        20,
        225,
        66,
        207,
        20,
        225,
        66,
        0,
        0,
        17,
        67,
        0,
        0,
        145,
        66,
        0,
        0,
        17,
        67,
        4,
        97,
        214,
        1,
        66,
        0,
        0,
        17,
        67,
        0,
        0,
        0,
        0,
        207,
        20,
        225,
        66,
        0,
        0,
        0,
        0,
        0,
        0,
        145,
        66,
        4,
        0,
        0,
        0,
        0,
        97,
        214,
        1,
        66,
        97,
        214,
        1,
        66,
        0,
        0,
        0,
        0,
        0,
        0,
        145,
        66,
        0,
        0,
        0,
        0,
        4,
        207,
        20,
        225,
        66,
        0,
        0,
        0,
        0,
        0,
        0,
        17,
        67,
        97,
        214,
        1,
        66,
        0,
        0,
        17,
        67,
        0,
        0,
        145,
        66,
        0,
      ],
    },
  ],
  "nodeChanges": Array [
    Object {
      "blendMode": "PASS_THROUGH",
      "guid": Object {
        "localID": 0,
        "sessionID": 0,
      },
      "mask": false,
      "maskIsOutline": false,
      "name": "Document",
      "opacity": 1,
      "phase": "CREATED",
      "transform": Object {
        "m00": 1,
        "m01": 0,
        "m02": 0,
        "m10": 0,
        "m11": 1,
        "m12": 0,
      },
      "type": "DOCUMENT",
      "visible": true,
    },
    Object {
      "backgroundColor": Object {
        "a": 1,
        "b": 0.9607843160629272,
        "g": 0.9607843160629272,
        "r": 0.9607843160629272,
      },
      "backgroundEnabled": true,
      "backgroundOpacity": 1,
      "blendMode": "PASS_THROUGH",
      "guid": Object {
        "localID": 1,
        "sessionID": 0,
      },
      "mask": false,
      "maskIsOutline": false,
      "name": "Page 1",
      "opacity": 1,
      "parentIndex": Object {
        "guid": Object {
          "localID": 0,
          "sessionID": 0,
        },
        "position": "!",
      },
      "phase": "CREATED",
      "transform": Object {
        "m00": 1,
        "m01": 0,
        "m02": 0,
        "m10": 0,
        "m11": 1,
        "m12": 0,
      },
      "type": "CANVAS",
      "visible": true,
    },
    Object {
      "blendMode": "PASS_THROUGH",
      "guid": Object {
        "localID": 2,
        "sessionID": 0,
      },
      "internalOnly": true,
      "mask": false,
      "maskIsOutline": false,
      "name": "Internal Only Canvas",
      "opacity": 1,
      "parentIndex": Object {
        "guid": Object {
          "localID": 0,
          "sessionID": 0,
        },
        "position": "~",
      },
      "phase": "CREATED",
      "transform": Object {
        "m00": 1,
        "m01": 0,
        "m02": 0,
        "m10": 0,
        "m11": 1,
        "m12": 0,
      },
      "type": "CANVAS",
      "visible": false,
    },
    Object {
      "arcData": Object {
        "endingAngle": 6.2831854820251465,
        "innerRadius": 0,
        "startingAngle": 0,
      },
      "blendMode": "PASS_THROUGH",
      "cornerRadius": 0,
      "cornerSmoothing": 0,
      "dashPattern": Array [],
      "fillGeometry": Array [
        Object {
          "commandsBlob": 0,
          "styleID": 0,
          "windingRule": "NONZERO",
        },
      ],
      "fillPaints": Array [
        Object {
          "blendMode": "NORMAL",
          "color": Object {
            "a": 1,
            "b": 0.8500000238418579,
            "g": 0.6204999685287476,
            "r": 0.53125,
          },
          "opacity": 1,
          "type": "SOLID",
          "visible": true,
        },
      ],
      "guid": Object {
        "localID": 2,
        "sessionID": 2,
      },
      "handleMirroring": "NONE",
      "horizontalConstraint": "MIN",
      "locked": false,
      "mask": false,
      "maskIsOutline": false,
      "miterLimit": 4,
      "name": "Ellipse 1",
      "opacity": 1,
      "parentIndex": Object {
        "guid": Object {
          "localID": 1,
          "sessionID": 0,
        },
        "position": "!",
      },
      "phase": "CREATED",
      "proportionsConstrained": true,
      "size": Object {
        "x": 145,
        "y": 145,
      },
      "strokeAlign": "INSIDE",
      "strokeCap": "NONE",
      "strokeGeometry": Array [],
      "strokeJoin": "MITER",
      "strokePaints": Array [],
      "strokeWeight": 1,
      "transform": Object {
        "m00": 1,
        "m01": 0,
        "m02": 0,
        "m10": 0,
        "m11": 1,
        "m12": 0,
      },
      "type": "ELLIPSE",
      "verticalConstraint": "MIN",
      "visible": true,
    },
  ],
  "sessionID": 0,
  "type": "NODE_CHANGES",
}
`;
